### ISSUE TITLE
Using CUDNN gives bad results.

### ISSUE THREAD

**Issue by isabeaups** [pytorch/pytorch#4845]
I have a next step prediction model on times series which is simply a GRU with a fully-connected layer on top of it. When I train it using CPU after 50 epochs I get a loss of 0.10 but when I train it with GPU the loss is 0.15 after 50 epochs. Doing more epochs doesnt really lower the losses in either cases.

I have tried changing the random seeds for both data and model, and these results are independent of the random seeds.

What does solve the problem is not using CUDNN:
If I add the line
`torch.backends.cudnn.enabled = False`
then training with GPU gives as good results as training with CPU.

This might be related to the issue that other people have that training with PyTorch gives worse results than Tensorflow: https://discuss.pytorch.org/t/the-same-model-produces-worse-results-on-pytorch-than-on-tensorflow/5380

I have:
Linux
Python 3.6.2
PyTorch 0.3.0
CUDNN_MAJOR 7
CUDNN_MINOR 0
CUDNN_PATCHLEVEL 5

GPU: Titan X Maxwell

**Comment by SsnL** [comment]
Could you share a bit more of your model, e.g. a script?

**Comment by isabeaups** [comment]
It is simply a GRU with a fully-connected output. 

    class SimpleRNN(Module):
        '''
        Simple RNN with teacher forcing option.
        '''
        def __init__(self, input_size, hidden_size, output_size=1, control_vec_size=0, weight_normalize=False):
            'Known inputs come last'
            super(SimpleRNN, self).__init__()
            self.input_size = input_size
            self.hidden_size = hidden_size
            self.output_size = output_size
            self.control_vec_size = control_vec_size
            self.weight_normalize = weight_normalize


            self.rnn = GRU(input_size=input_size, hidden_size=hidden_size, num_layers=1)
            # Initialize to encourage learning of the slow dynamics
            for name, param in self.rnn.named_parameters():
                if 'bias' in name:

                    nn.init.normal(param)
                elif 'weight' in name:
                    nn.init.xavier_normal(param)

            self.fc = Linear(hidden_size, output_size)

            if weight_normalize:
                rnn_names = [n for n, v in self.rnn.named_parameters()]
                fc_names = [n for n,v in self.fc.named_parameters()]
                for n in rnn_names:
                    if 'weight' in n:
                        weight_norm(self.rnn, n)
                for n in fc_names:
                    if 'weight' in n:
                        weight_norm(self.fc, n)


        def step(self, input, hidden=None):
            # input is (batch, inp_size), hidden is tuple
            rnn_output, hidden = self.rnn(input.unsqueeze(0), hidden) # rnn_output is (1, batch, hidden_size)
            output = self.fc(rnn_output.squeeze(0)) # (batch, out_size)
            return output, hidden

        def forward(self, inputs, hidden=None, gen_poisson_lambda=0):
            n_ahead = 1
            n_steps = len(inputs)
            n_batch = inputs.size(1)
            outputs = Variable(torch.zeros(n_steps-n_ahead, n_batch, self.output_size))
            left_to_gen = 0
            for t in range(n_steps-1):
                if left_to_gen == 0:
                    inp = inputs[t]
                    left_to_gen = np.random.poisson(gen_poisson_lambda)
                else:
                    inp = torch.cat([output[:, :-self.control_vec_size], inputs[t, :, -self.control_vec_size:]], dim=1)
                    left_to_gen -= 1
                output, hidden = self.step(inp, hidden)
                outputs[t] = output
            return outputs, hidden

**Comment by SsnL** [comment]
This is interesting. I tried w/ and w/o cudnn on simple inputs and got identical gradients. Could you try not passing in each time step at a time but activating the entire sequence all together? cudnn is probably more optimized to that.

**Comment by ngimel** [comment]
It also might be interaction between weight norm and parameter flattening for cudnn. Can you try your network without weight norm and see if with cudnn and without cudnn produce similar results?

**Comment by isabeaups** [comment]
@ngimel Yes! When I turned off weight normalization it worked also with CUDNN. So it does seem to be the conjunction of CUDNN and weight normalization which creates the problem.

**Comment by mruberry** [comment]
Closing this due to age. If this issue still occurs with a recent version of PyTorch and cuDNN please update and reopen.
